registry/registry-polaris/src/main/java/com/alipay/sofa/rpc/registry/polaris/PolarisRegistry.java

```diff
@@ -0,0 +1,373 @@
+                }
+                if (LOGGER.isInfoEnabled(appName)) {
+                    LOGGER.infoWithApp(appName,
+                            LogCodes.getLog(LogCodes.INFO_ROUTE_REGISTRY_PUB_OVER, config.getInterfaceId()));
+                }
+            }
+        } catch (
+                SofaRpcRuntimeException e) {
+            throw e;
+        } catch (Exception e) {
+            throw new SofaRpcRuntimeException(LogCodes.getLog(LogCodes.ERROR_REG_PROVIDER, "polarisRegistry", config.buildKey()), e);
+        }
+        if (EventBus.isEnable(ProviderPubEvent.class)) {
+            ProviderPubEvent event = new ProviderPubEvent(config);
+            EventBus.post(event);
+        }
+
+    }
+
+    //convert ProviderConfig to polaris registerRequest
+    public List<InstanceRegisterRequest> buildPolarisRegister(ProviderConfig config) {
+
+        List<ServerConfig> serverConfigs = config.getServer();
+        if (CommonUtils.isEmpty(serverConfigs)) {
+            return Collections.emptyList();
+        }
+
+        List<InstanceRegisterRequest> requestList = new ArrayList<>();
+        for (ServerConfig serverConfig : serverConfigs) {
+            InstanceRegisterRequest request = new InstanceRegisterRequest();
+            request.setNamespace(buildNameSpace(config.getAppName()));
+            request.setService(buildServiceName(config));
+            request.setHost(getServerHost(serverConfig));
+            request.setPort(serverConfig.getPort());
+            request.setPriority(config.getPriority());
+            request.setProtocol(serverConfig.getProtocol());
+            request.setWeight(config.getWeight());
+            request.setTimeoutMs(config.getTimeout());
+            request.setVersion(config.getVersion());
+            request.setTtl(properties.getHealthCheckTTL());
+            Map<String, String> metaData = convertProviderToMap(config, serverConfig);
+            checkAndDelNull(metaData);
+            request.setMetadata(metaData);
+            requestList.add(request);
+        }
+        return requestList;
+    }
+
+    private String buildNameSpace(String appName) {
+        return null == appName ? "sofa-default" : appName;
+    }
+
+    private void checkAndDelNull(Map<String, String> metaData) {
+        metaData.entrySet().removeIf((e) -> e.getValue() == null);
+    }
+
+    private void registerPolarisService(ProviderConfig config, InstanceRegisterRequest service) {
+        providerAPI.register(service);
+        if (service.getTtl() != null) {
+            ScheduledFuture<?> scheduledFuture =
+                    heartbeatExecutor.scheduleAtFixedRate(
+                            () -> heartbeatPolaris(service),
+                            0, properties.getHeartbeatInterval(), TimeUnit.MILLISECONDS);
+            ScheduledFuture oldFuture = heartbeatFutures.remove(buildUniqueName(config,service.getProtocol()));
+            if (oldFuture != null) {
+                oldFuture.cancel(true);
```

Comment:
Can use PolarisHeartbeatTask implement `com.alipay.sofa.rpc.base.Destroyable` to replace future to optimize. Future cancel depends on thread interruption.

Reply:
Is the current logic wrong? I don't quite understand it, can you explain it in detail?

Reply:
May have this problem like this `https://blog.csdn.net/stephen8341/article/details/50433656`.

Reply:
This is not a big problem. Even if the task will be executed this time, the next and subsequent scheduled tasks will be cancelled.